cdap-tms/src/main/java/co/cask/cdap/messaging/server/FetchHandler.java

```diff
@@ -240,11 +251,14 @@ public void finished() throws Exception {
     public void handleError(@Nullable Throwable cause) {
       iterator.close();
       // Since response header is already sent, there is nothing we can send back to client. Simply log the failure
-      if (cause instanceof SocketException) {
+      if (cause instanceof SocketException
+        || cause instanceof ClosedChannelException
+        || (cause instanceof IOException && KNOWN_IO_EXCEPTION_MESSAGES.contains(cause.getMessage()))) {
         // This can easily caused by client close connection prematurely. Don't want to flood the log.
-        LOG.debug("Socket exception raised when sending messages back to client", cause);
+        LOG.trace("Connection closed by client prematurely while sending messages back to client", cause);
       } else {
-        LOG.warn("Exception raised when sending messages back to client", cause);
+        // Use sampling logger to log to avoid flooding the log if there is any systematic failure
+        SAMPLING_LOG.warn("Exception raised when sending messages back to client", cause);
```

Comment:
would you still log it as trace(). The sampling logger would suppress some messages completely...

Reply:
Added a trace

Reply:
Do you want to use the same message as above?

Reply:
ah they are different